scripts/news/utils.py

```diff
@@ -0,0 +1,250 @@
+        version_data=version_data,
+        metadata=metadata,
+    )
+
+    done = []
+    for line in res.split("\n"):
+        if wrap:
+            done.append(
+                textwrap.fill(
+                    line,
+                    width=79,
+                    subsequent_indent=" ",
+                    break_long_words=False,
+                    break_on_hyphens=False,
+                )
+            )
+        else:
+            done.append(line)
+
+    return "\n".join(done).rstrip() + "\n"
+
+
+def _uniquify(path: Path, sep: str = ".") -> Path:
+    """Expands name portion of filename with numeric '.(x)' suffix to return a unique path."""
+
+    def name_sequence() -> Generator[str, Any, None]:
+        count = itertools.count(start=1)
+        yield ""
+        while True:
+            yield "{s}{n:d}".format(s=sep, n=next(count))
+
+    orig = tempfile._name_sequence
+    with tempfile._once_lock:
+        tempfile._name_sequence = name_sequence()
+        dirname, filename, ext = path.parent, path.stem, path.suffix
+        fd, filename = tempfile.mkstemp(dir=dirname, prefix=filename, suffix=ext)
+        tempfile._name_sequence = orig
+
+    return Path(filename)
+
+
+def save_news_fragment(ctx: click.Context, gh_pr: int, news_entry: str, news_type: str) -> None:
+    """Save received changelog data to a news file."""
+    date = datetime.now(timezone.utc).strftime("%Y-%m-%d")
+    path = Path(REPO_ROOT, f"news/next/{news_type}/{date}.pr-{gh_pr}.md")
+    if not path.parents[1].exists():
+        err(NO_NEWS_PATH_ERROR, fg="blue")
+        ctx.exit(1)
+    elif not path.parents[0].exists():
+        make_news_type_dir = click.confirm(
+            f"Should I make the new type DIR for the news type at {path.parents[0]}"
+        )
+        if make_news_type_dir:
+            path.parents[0].mkdir(exist_ok=True)
+    elif path.exists():
+        # a changelog for that PR already
+        one_more = click.prompt(
+            f"A changelog for this PR already exists "
+            f"at {Path(os.path.relpath(path, start=Path.cwd()))}, "
+            f"do you want to make another one?"
+        )
+        if one_more:
+            path = _uniquify(path)
+        else:
+            ctx.exit(1)
+
+    text = str(news_entry)
+    with open(path, "wt", encoding="utf-8") as file:
+        file.write(text)
+
+    out(
+        f"All done! ✨ 🍰 ✨ Created news fragment at {Path(os.path.relpath(path, start=Path.cwd()))}"
+        "\nYou are now ready for commit the changelog!"
+    )
+
+
+def validate_pull_request_number(
+    ctx: click.Context, _param: click.Parameter, value: Optional[int]
+) -> Optional[int]:
+    """Check if the given pull request number exists on the github repository."""
+    r = requests.get(PR_ENDPOINT.format(number=value))
+    if r.status_code == 403:
+        if r.headers.get("X-RateLimit-Remaining") == "0":
+            err(f"{ERROR_MSG_PREFIX} Ratelimit reached, please retry in a few minutes.")
```

Comment:
how much might this be possible? Depending on the possibility, should try from the start with the GITHUB_TOKEN environment variable if it exists. This would also be accessible in the workflow, which would mean less possible ratelimits. \:D

Reply:
Never, the rate limits are 5k per hour which we aren't gonna hit even if we use it as a workflow.